core/cache.py

```
@@ -0,0 +1,152 @@
+        """Retrieve value from cache with error handling."""
+        try:
+            async with self._lock:
+                value = self._cache.get(key)
+                if value is not None:
+                    logger.debug(f"Cache hit for key: {key}")
+                else:
+                    logger.debug(f"Cache miss for key: {key}")
+                return value
+        except Exception as e:
+            logger.error(f"Error retrieving from cache for key {key}: {e}")
+            raise CacheError(f"Failed to retrieve from cache: {e}")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)